plugin/action/throttle/limiters_map.go

```diff
@@ -0,0 +1,209 @@
+package throttle
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis"
+	"github.com/ozontech/file.d/pipeline"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// interface with only necessary functions of the original redis.Client
+type redisClient interface {
+	IncrBy(key string, value int64) *redis.IntCmd
+	Expire(key string, expiration time.Duration) *redis.BoolCmd
+	SetNX(key string, value any, expiration time.Duration) *redis.BoolCmd
+	Get(key string) *redis.StringCmd
+	Ping() *redis.StatusCmd
+}
+
+type limiter interface {
+	isAllowed(event *pipeline.Event, ts time.Time) bool
+	sync()
+}
+
+// limiterWithGen is a wrapper for the limiter interface with added generation field.
+type limiterWithGen struct {
+	limiter
+	gen *atomic.Int64
+}
+
+func newLimiterWithGen(lim limiter, gen int64) *limiterWithGen {
+	return &limiterWithGen{
+		limiter: lim,
+		gen:     atomic.NewInt64(gen),
+	}
+}
+
+// limiterConfig configuration for creation of new limiters.
+type limiterConfig struct {
+	ctx               context.Context
+	backend           string
+	redisClient       redisClient
+	pipeline          string
+	throttleField     string
+	bucketInterval    time.Duration
+	bucketsCount      int
+	limiterValueField string
+}
+
+// limitersMapConfig configuration of limiters map.
+type limitersMapConfig struct {
+	limitersExpiration time.Duration
+	logger             *zap.SugaredLogger
+	syncWorkerCount    int
+	syncInterval       time.Duration
+	limiterCfg         *limiterConfig
+}
+
+// limitersMap is auxiliary type for storing the map of strings to limiters with additional info for cleanup
+// and thread safe map operations handling.
+type limitersMap struct {
+	lims            map[string]*limiterWithGen
+	mu              *sync.RWMutex
+	limiterBuf      []byte
+	curGen          int64
+	limitersExp     int64
+	logger          *zap.SugaredLogger
+	syncWorkerCount int
+	syncInterval    time.Duration
+
+	limiterCfg *limiterConfig
+}
+
+func newLimitersMap(lmCfg limitersMapConfig, redisOpts *redis.Options) *limitersMap {
+	nowTs := time.Now().UnixMicro()
+	lm := &limitersMap{
+		lims:            make(map[string]*limiterWithGen),
+		mu:              &sync.RWMutex{},
+		limiterBuf:      make([]byte, 0),
+		curGen:          nowTs,
+		limitersExp:     lmCfg.limitersExpiration.Microseconds(),
+		logger:          lmCfg.logger,
+		syncWorkerCount: lmCfg.syncWorkerCount,
+		syncInterval:    lmCfg.syncInterval,
+
+		limiterCfg: lmCfg.limiterCfg,
+	}
+	if redisOpts != nil {
+		client := redis.NewClient(redisOpts)
+		if pingResp := client.Ping(); pingResp.Err() != nil {
+			lm.logger.Fatalf("can't ping redis: %s", pingResp.Err())
+		}
+		lm.limiterCfg.redisClient = client
+	}
+	return lm
+}
+
+func (l *limitersMap) syncWorker(ctx context.Context, jobCh <-chan limiter, wg *sync.WaitGroup) {
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
Deadlock is possible here when we are `wg.Add(1)`, but workers exit here

Reply:
done